Maas docs

nottrobin opened the pull request:
The new docs section - maas.io/docs - to replace docs.maas.io.

## QA

Get hold of the "Custom Search API key" from [the Google cloud API credentials page](https://console.cloud.google.com/apis/credentials?project=ubuntu-search-1530889417216) - if you don't have access to that page, ask @nottrobin or someone for access.

Then, run the site with:

``` bash
./run --env SEARCH_API_KEY={the-api-key}
```

Now go to http://0.0.0.0:8006/docs, check it looks good, click around, check links work etc.

webteam-app commented:
Starting demo at: https://maas-io-canonical-web-and-design-pr-346.run.demo.haus/

webteam-app commented:
Starting demo at: https://maas-io-canonical-web-and-design-pr-346.run.demo.haus/

lilyvidenova commented:
Just like snapcraft's one - no extra scroll bar, please! :)

lilyvidenova commented:
Hmmm, on the PR above all of the headings that are also links in the navigation are blue, could we have them be the same as the other navigation headings? (with the active/link styling on hover?)

lilyvidenova commented:
Not sure if part of this piece of work, but can we persist the search bar on the search results page - the way that snapcraft does it?

lilyvidenova commented:
And one last question - where can a user find the docs for older versions of maas?

nottrobin commented:
> @lilyvidenova:
> Just like snapcraft's one - no extra scroll bar, please! :)
> Hmmm, on the PR above all of the headings that are also links in the navigation are blue, could we have them be the same as the other navigation headings? (with the active/link styling on hover?)
> Not sure if part of this piece of work, but can we persist the search bar on the search results page - the way that snapcraft does it?

Yep, will do all those

> @lilyvidenova: And one last question - where can a user find the docs for older versions of maas?

Going forward, changes between versions should be documented inline - as in a document on "add machines" it might say "Note: X feature is only available from version 2.7 onwards" or "Note: Coming in 2.8 ...".

But for older versions of MAAS (prior to 2.6), the docs will remain available at  old-docs.maas.io - e.g. https://old-docs.maas.io/2.5/en/, and existing links to old versions of documentation will forward there.

Currently we don't have a plan to explicitly link to `old-docs` from anywhere in the new documentation platform - but maybe we should. We could either do this by mentioning it on the homepage, or linking to it in the navigation, or something. What do you think?

lilyvidenova commented:
> Currently we don't have a plan to explicitly link to `old-docs` from anywhere in the new documentation platform - but maybe we should. We could either do this by mentioning it on the homepage, or linking to it in the navigation, or something. What do you think?

@nottrobin  I think we definitely need to link to the `old-docs` as a lot of users are still on older versions of MAAS.
Let's put a link in the:
-  "Getting support" section, with the title "Pre-2.5 documentation" on the Home page
-  Under "Troubleshoot" in the left-hand-nav, with the same title.

Another minor thing I just noticed:
![image](https://user-images.githubusercontent.com/17980375/61372886-333e7900-a890-11e9-8285-e5a9d3d44a68.png)

lilyvidenova commented:
I updated the things above in discourse.

lilyvidenova commented:
@nottrobin 
- Can we use the same Next and Previous buttons on the results page as snapcraft? 
<img width="1680" alt="Screenshot 2019-07-23 at 14 57 53" src="https://user-images.githubusercontent.com/17980375/61718308-a21d4580-ad5a-11e9-9d0b-97d8c64c690e.png">

- The API documentation link doesn't work

Other than that - looks great, thanks Robin! :)

tbille commented on webapp/search.py:
```diff
@@ -0,0 +1,45 @@
+from canonicalwebteam.http import CachedSession
```
I think you forgot to remove this file. (With the new search module)

nottrobin replied:
Thanks! Removed.

tbille commented on templates/docs/document.html:
```diff
@@ -0,0 +1,74 @@
+{% block content %}
+  <section id="search-docs" class="p-strip--image is-shallow" style="background-image: url('https://assets.ubuntu.com/v1/e54487e2-maas-docs-suru.png')">
+    <div class="row">
+      <form class="p-search-box u-no-margin--bottom" style="box-shadow: none" action="/docs/search">
```
on links I try to use `url_for('name of the function')` to not be dependent of the urls in case there is a change. 

`url_for` will get the defined url in the flask routes. https://flask.palletsprojects.com/en/1.1.x/api/?highlight=url_for#flask.url_for

nottrobin replied:
I'm aware of `url_for`. I prefer not to use it. The reason is that to me its usage us about where you want your locus of control to be. If you expect the important people trying to understand you app to be reading Python and want to avoid reading HTML, then they would rather be able to find and change all the URLs in `app.py` or `views.py` or `urls.py` or whatever.

However, if they are more likely to be reading HTML, then having to work out how `url_for` generates a URL and where its stored is hard for them.

I think it's quite clear in our case that it's much more useful for the majority of the team for the HTML to be direct than for everything to be controlled through Python.

tbille commented on webapp/app.py:
```diff
@@ -2,25 +2,60 @@
+
+# Search
+app.add_url_rule(
+    "/docs/search",
```
just to be picky :P

maybe `DOCS_URL_PREFIX + /search` ?

nottrobin replied:
Why does `DOCS_URL_PREFIX` exist?

nottrobin replied:
Some sources for my point of view:

https://en.wikipedia.org/wiki/Convention_over_configuration
https://medium.com/@rdsubhas/10-modern-software-engineering-mistakes-bc67fbef4fc8#97ef

nottrobin replied:
Where are you seeing `DOCS_URL_PREFIX`? `rg DOCS_URL_PREFIX` returns nothing for me.

tbille replied:
https://github.com/canonical-web-and-design/maas.io/pull/346/files#diff-bed92ddcb62003ca6b223dfa86df23b3R20

(Line 20 of this file)

nottrobin replied:
Sorry I did find it. Yeah I can see that since it existed,I should use it. So I removed it - problem solved =)

tbille replied:
I don't really mind not using the prefix in this way. But then all those variables that are defined could be removed since only used in the DiscourseDocs class

nottrobin replied:
Yep, thanks for bringing them to my attention. I also removed all the redundant search config =)

tbille commented on webapp/app.py:
```diff
@@ -35,3 +70,14 @@ def internal_error(error):
     return dict(get_rss_feed=get_rss_feed)
+
+
+@app.route("/docs/api")
```
same here

`DOCS_URL_PREFIX + /api` ?

nottrobin replied:
Removed `DOCS_URL_PREFIX`

tbille commented on webapp/search.py:
```diff
@@ -0,0 +1,39 @@
+class NoAPIKeyError(Exception):
```
the file is still there :P

nottrobin replied:
lol

nottrobin replied:
Removed. Thanks!

tbille commented on webapp/app.py:
```diff
@@ -35,3 +59,14 @@ def internal_error(error):
+    Show the static api page
+    """
+
+    parser = DocParser(discourse_api, DOCS_TOPIC_ID, DOCS_URL_PREFIX)
```
actually this was the reason why you had a variable sorry I missed this

nottrobin replied:
yeah but this was the thing we solved with version `0.6.0` of discourse-docs, so I'm just upgrading.

nottrobin commented:
@lilyvidenova I've fixed the pagination issue. I've also updated the navigation content in the forum so "API reference" links to https://maas.io/docs/api. This won't work now, because maas.io/docs redirects to docs.maas.io - but as soon as this goes live this will work - you can see by visiting https://maas-io-canonical-web-and-design-pr-346.run.demo.haus/api

lilyvidenova commented:
@nottrobin awesome, thanks, looks great! 👍

nottrobin commented on templates/includes/footer.html:
```diff
@@ -16,7 +16,7 @@ <h3 class="p-heading--four">MAAS</h3>
                 </li>
                 <li class="p-list__item">
-                    <a class="p-link--external" href="https://docs.maas.io">Docs</a>
+                    <a class="p-link--strong" href="/docs">Docs</a>
```
Cool I've removed it in these two cases, but there are still a lot of them in the `index.html` which I'm not going to touch in this PR.

``` bash
$ rg p-link--strong
index.html
50:            <a class="p-link--strong p-heading--five" aria-label="Learn more about automatic discovery and registration of every device on the network" href="/tour#automation" onClick="dataLayer.push({'event' : 'GAEvent', 'eventCategory' : 'Feature', 'eventAction' : 'Click', 'eventLabel' : 'Automation' }); dataLayer.push({'event' : 'GAEvent', 'eventCategory' : 'Interested in feature', 'eventAction' : 'Click', 'eventLabel' : 'Learn more about device discovery and subnet mapping' });">Automation&nbsp;&rsaquo;</a>
71:            <a class="p-link--strong p-heading--five" href="/tour#fast-deployment" onClick="dataLayer.push({'event' : 'GAEvent', 'eventCategory' : 'Feature', 'eventAction' : 'Click', 'eventLabel' : 'Learn more about images' }); dataLayer.push({'event' : 'GAEvent', 'eventCategory' : 'Feature', 'eventAction' : 'Click', 'eventLabel' : 'Fast deployment' });">Speed&nbsp;&rsaquo;</a>
83:            <a class="p-link--strong p-heading--five" href="/how-it-works#node-lifecycle" onClick="dataLayer.push({'event' : 'GAEvent', 'eventCategory' : 'Feature', 'eventAction' : 'Click', 'eventLabel' : 'Chassis management' });">Inventory&nbsp;&rsaquo;</a>
94:            <a class="p-link--strong p-heading--five" href="/tour#configure-servers" onClick="dataLayer.push({'event' : 'GAEvent', 'eventCategory' : 'Feature', 'eventAction' : 'Click', 'eventLabel' : 'Learn more about storage' }); dataLayer.push({'event' : 'GAEvent', 'eventCategory' : 'Feature', 'eventAction' : 'Click', 'eventLabel' : 'Machine configuration' });">Storage layouts&nbsp;&rsaquo;</a>
108:            <a class="p-link--strong p-heading--five" href="/tour#configure-servers" onClick="dataLayer.push({'event' : 'GAEvent', 'eventCategory' : 'Feature', 'eventAction' : 'Click', 'eventLabel' : 'Learn more about storage' }); dataLayer.push({'event' : 'GAEvent', 'eventCategory' : 'Feature', 'eventAction' : 'Click', 'eventLabel' : 'Machine configuration' });">IPAM, DHCP, DNS&nbsp;&rsaquo;</a>
121:            <a class="p-link--strong p-heading--five" href="/tour#hardware-testing" onClick="dataLayer.push({'event' : 'GAEvent', 'eventCategory' : 'Feature', 'eventAction' : 'Click', 'eventLabel' : 'Hardware testing' });">Hardware testing&nbsp;&rsaquo;</a>
145:            <a class="p-link--strong p-heading--five" href="/tour#manage-network" onClick="dataLayer.push({'event' : 'GAEvent', 'eventCategory' : 'Feature', 'eventAction' : 'Click', 'eventLabel' : 'Network management' });">Network monitoring&nbsp;&rsaquo;</a>
```

nottrobin commented:
Thanks @barrymcgee. I've fixed the weird layout below the table ("With this complete..."), and I've increased the spacing around tables and images. Could you take another look?
